apis/v1alpha2/tlsroute_types.go

```diff
@@ -49,26 +49,35 @@ type TLSRouteSpec struct {
 	CommonRouteSpec `json:",inline"`
 
 	// Hostnames defines a set of SNI names that should match against the
-	// SNI attribute of TLS ClientHello message in TLS handshake.
-	//
-	// SNI can be "precise" which is a domain name without the terminating
-	// dot of a network host (e.g. "foo.example.com") or "wildcard", which is
-	// a domain name prefixed with a single wildcard label (e.g. `*.example.com`).
-	// The wildcard character `*` must appear by itself as the first DNS label
-	// and matches only a single label. You cannot have a wildcard label by
-	// itself (e.g. Host == `*`).
-	//
-	// Requests will be matched against the SNI attribute in the following
-	// order:
-	//
-	// 1. If SNI is precise, the request matches this Route if the SNI in
-	//    ClientHello is equal to one of the defined SNIs.
-	// 2. If SNI is a wildcard, then the request matches this Route if the
-	//    SNI is to equal to the suffix (removing the first label) of the
-	//    wildcard.
-	// 3. If SNIs are unspecified, all requests associated with the gateway TLS
-	//    listener will match. This can be used to define a default backend
-	//    for a TLS listener.
+	// SNI attribute of TLS ClientHello message in TLS handshake. This matches
+	// the RFC 1123 definition of a hostname with 2 notable exceptions:
+	//
+	// 1. IPs are not allowed.
+	// 2. A hostname may be prefixed with a wildcard label (`*.`). The wildcard
+	//    label must appear by itself as the first label.
+	//
+	// If a hostname is specified by both the Listener and TLSRoute, there
+	// must be at least one intersecting hostname for the TLSRoute to be
+	// attached to the Listener. For example:
+	//
+	// * A Listener with `test.example.com` as the hostname matches TLSRoutes
+	//   that have either not specified any hostnames, or have specified at
+	//   least one of `test.example.com` or `*.example.com`.
+	// * A Listener with `*.example.com` as the hostname matches TLSRoutes
+	//   that have either not specified any hostnames or have specified at least
+	//   one hostname that matches the Listener hostname. For example,
+	//   `test.example.com` and `*.example.com` would both match. On the other
+	//   hand, `example.com` and `test.example.net` would not match.
+	//
+	// If both the Listener and TLSRoute have specified hostnames, any
+	// TLSRoute hostnames that do not match the Listener hostname MUST be
+	// ignored. For example, if a Listener specified `*.example.com`, and the
+	// TLSRoute specified `test.example.com` and `test.example.net`,
+	// `test.example.net` must not be considered for a match.
+	//
+	// If hostnames do not match with the criteria above, then the TLSRoute is
+	// not accepted, and the implementation must raise an 'Accepted' Condition
+	// with a status of `False` for the target Listener(s).
```

Comment:
This is a bit confusing when not using `sectionName` in parentRefs. Since the target is not a listener

Reply:
Yeah, this comes down to the behavior with partial Listener acceptance, I think if we mandate "all", and have a ResolvedRefs in the event not all match, then this covers the sectionName case.

Reply:
I agree that we need a better way to handle status when `sectionName` is not specified. 2 options I've been thinking of:

1. A standard/reserved listener name such as "all" that is used in status to cover failed attachment to the Gateway as a whole.
2. A new set of condition types for Gateway status specifically for this purpose. (We could potentially reuse the existing listener condition types).

Reply:
I think option 2 seems better - if no listeners are valid, that seems very relevant for top-level status, but I would have thought we had that covered already somehow.

Reply:
https://github.com/kubernetes-sigs/gateway-api/issues/849